### PR TITLE
[SPARK-42238][SQL] Introduce new error class: `INCOMPATIBLE_JOIN_TYPES`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to introduce new error class `INCOMPATIBLE_JOIN_TYPES` to improve the error message for incompatible join type usage.

### Why are the changes needed?

The existing error classes `LATERAL_NATURAL_JOIN` and `NATURAL_CROSS_JOIN` are not logically belong under `UNSUPPORTED_FEATURE`, and their error message is not very clear to understand for end-users.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Updated UTs.

### CODE REVIEW

**Comment by itholic** [comment]
FYI @srielau @MaxGekk @cloud-fan

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -586,6 +586,12 @@
       "Detected an incompatible DataSourceRegister. Please remove the incompatible library from classpath or upgrade it. Error: <message>"
     ]
   },
+  "INCOMPATIBLE_JOIN_TYPES" : {
+    "message" : [
+      "The join types <joinType1> and <joinType2> are incompatible."
+    ],
+    "sqlState" : "42818"
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:593]: 42818 is about operators (e.g. /, * +) and unctions.
```suggestion
    "sqlState" : "42613"
```

Note I think there is nothing wrong with a LATERAL on the RHS of of one of these joins.
SELECT * FROM T LEFT JOIN S RIGHT JOIN LATERAL(SELECT T.c1) .... should be legal

**Review by srielau — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 2]
```diff
@@ -165,10 +158,13 @@ private[sql] object QueryParsingErrors extends QueryErrorsBase {
...
-  def naturalCrossJoinUnsupportedError(ctx: ParserRuleContext): Throwable = {
+  def incompatibleJoinTypesError(
+      joinType1: String, joinType2: String, ctx: ParserRuleContext): Throwable = {
     new ParseException(
-      errorClass = "UNSUPPORTED_FEATURE.NATURAL_CROSS_JOIN",
-      messageParameters = Map.empty,
+      errorClass = "INCOMPATIBLE_JOIN_TYPES",
+      messageParameters = Map(
+        "joinType1" -> toSQLType(joinType1),
+        "joinType2" -> toSQLType(joinType2)),
```
[the remark this answers is not in the document]
Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala]: Updated, thanks!

**Comment by MaxGekk** [comment]
@itholic Please, resolve conflicts.

**Comment by itholic** [comment]
Resolved!

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.4.
Thank you, @itholic and @cloud-fan @srielau for review.

**Pull request closed without merging**